src/main/kotlin/org/jitsi/nlj/transform/node/Node.kt

```diff
@@ -368,26 +395,26 @@ abstract class PredicateFilterNode(
 /**
  * A [Node] which observes packets, but makes no modifications.
  */
-abstract class ObserverNode(name: String) : TransformerNode(name) {
+abstract class ObserverNode(name: String) : NeverDiscardNode(name) {
 
     protected abstract fun observe(packetInfo: PacketInfo)
 
-    override fun transform(packetInfo: PacketInfo): PacketInfo? {
+    override fun handlePacket(packetInfo: PacketInfo) {
         observe(packetInfo)
-        return packetInfo
     }
 }
 
 /**
  * A node which consumes all packets (i.e. does something with them, but does not forward them to another node).
  */
-abstract class ConsumerNode(name: String) : TransformerNode(name) {
+abstract class ConsumerNode(name: String) : StatsKeepingNode(name) {
 
     protected abstract fun consume(packetInfo: PacketInfo)
 
-    override fun transform(packetInfo: PacketInfo): PacketInfo? {
+    override fun doProcessPacket(packetInfo: PacketInfo) {
+        // TODO(before merging): should we call doneProcessing here?
+        //  Before, or after calling consume?
```

Comment:
Am I confused?  Where does `ConsumerNode` discard now?

Reply:
`ConsumerNode` doesn't discard at all...the 'consumer' consumes the packet (meaning it continues to use it)

Reply:
I don't think if it should discard: it's previous parent `TransformerNode` didn't.

Every direct child of `StatsKeepingNode` should call `doneProcessing` after performing it's work, so that `StatsKeepingNode` can do it's job of keeping track of processing time (this requirement is why we made `StatsKeepingNode` a sealed class IIRC).

Reply:
```
    /**
     * Should be called by sub classes when they finish processing of the input packet, but before they call into any
     * other nodes, so that [Node] can keep track of its statistics.
     */
protected fun doneProcessing(packetInfo: PacketInfo?)
```

Reply:
Yes, I think my comment was unclear but I wasn't wondering about whether or not we should call `doneProcessing`, just _when_ it should be called.  Thinking about it now, I think preserving the old behavior (calling `doneProcessing` after `consume`) makes sense.

Reply:
To take a specific case, for the receive pipeline's `PacketHandlerWrapper`, this means it counts the time spent by the Conference `sendOut` code, right? This should be useful for looking at my https://github.com/jitsi/jitsi-videobridge/pull/1062 .